File: frontends/configurator/www/model/config/entities/possible_value.js
import {Entities} from '../entities.js';
import {EntitiesHooks} from '../entities_hooks.js';
import {Node} from '../node.js';
import {Utils} from '../utils.js';

export class PossibleValue extends Node {
	static getProperties() {
		return {
			fieldModel: {type: Entities.FieldModel.name, back_reference: true},
			id: {type: 'string'},
			shortname: {type: 'object'},
			exportLabel: {type: 'string'},
			specify: {type: 'boolean'},
		};
	}

	constructor(values) {
		super();
		this.fieldModel = undefined;
		this.id = undefined;
		this.shortname = {};
		this.exportLabel = undefined;
		this.specify = false;
		EntitiesHooks?.CreateNode.call(this, values);
	}

	//shortname
	getLocalizedShortname(languages) {
		return Utils.getLocalizedField.call(this, 'shortname', languages);
	}
	getLocalizedLabel(languages) {
		return this.getLocalizedShortname(languages) || this.id;
	}
}

File: backend/src/main/java/ch/rodano/configuration/model/field/PossibleValue.java
package ch.rodano.configuration.model.field;

import java.util.Collection;
import java.util.Collections;
import java.util.Map;
import java.util.TreeMap;

import org.apache.commons.lang3.StringUtils;

import com.fasterxml.jackson.annotation.JsonBackReference;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonInclude.Include;
import com.fasterxml.jackson.annotation.JsonPropertyOrder;

import ch.rodano.configuration.model.common.Entity;
import ch.rodano.configuration.model.common.Node;
import ch.rodano.configuration.utils.DisplayableUtils;

@JsonInclude(Include.NON_NULL)
@JsonPropertyOrder(alphabetic = true)
public class PossibleValue implements Node {
	private static final long serialVersionUID = -10498351329495800L;

	private String id;
	private FieldModel fieldModel;

	private Map<String, String> shortname;

	private String exportLabel;

	private boolean specify;

	public PossibleValue() {
		shortname = new TreeMap<>();
	}

	public PossibleValue(final String id, final Map<String, String> shortname) {
		this.id = id;
		this.shortname = shortname;
	}

	public final String getId() {
		return id;
	}

	public final void setId(final String id) {
		this.id = id;
	}

	@JsonBackReference
	public final FieldModel getFieldModel() {
		return fieldModel;
	}

	@JsonBackReference
	public final void setFieldModel(final FieldModel fieldModel) {
		this.fieldModel = fieldModel;
	}

	public final String getExportLabel() {
		return exportLabel;
	}

	public final void setExportLabel(final String exportLabel) {
		this.exportLabel = exportLabel;
	}

	public final Map<String, String> getShortname() {
		return shortname;
	}

	public final void setShortname(final Map<String, String> shortname) {
		this.shortname = shortname;
	}

	public final boolean isSpecify() {
		return specify;
	}

	public final void setSpecify(final boolean specify) {
		this.specify = specify;
	}

	@JsonIgnore
	public String getLocalizedShortname(final String... languages) {
		return DisplayableUtils.getLocalizedMapWithDefault(shortname, "", languages);
	}

	@JsonIgnore
	public final String getDefaultLocalizedShortname() {
		return getLocalizedShortname(getFieldModel().getDatasetModel().getStudy().getDefaultLanguageId());
	}

	@JsonIgnore
	public final String getExportColumnLabel() {
		return StringUtils.defaultIfBlank(exportLabel, id);
	}

	@Override
	public final Entity getEntity() {
		return Entity.POSSIBLE_VALUE;
	}

	@Override
	public Collection<Node> getChildrenWithEntity(final Entity entity) {
		return Collections.emptyList();
	}
}
